modules/dex/src/lib.rs

```diff
@@ -18,41 +17,40 @@ use orml_traits::{MultiCurrency, MultiCurrencyExtended};
 use orml_utilities::with_transaction_result;
 use primitives::{Balance, CurrencyId};
 use sp_runtime::{
-	traits::{AccountIdConversion, CheckedAdd, CheckedMul, CheckedSub, One, Saturating, Zero},
+	traits::{AccountIdConversion, One, UniqueSaturatedInto, Zero},
 	DispatchError, DispatchResult, FixedPointNumber, ModuleId,
 };
-use sp_std::prelude::Vec;
-use support::{CDPTreasury, DEXManager, Price, Rate, Ratio};
+use sp_std::{prelude::*, vec};
+use support::{DEXManager, Price, Ratio};
 
 mod benchmarking;
 mod default_weight;
 mod mock;
-mod tests;
+// mod tests;
 
 pub trait WeightInfo {
 	fn add_liquidity() -> Weight;
 	fn withdraw_liquidity() -> Weight;
-	fn swap_currency() -> Weight;
+	fn swap_with_exact_supply() -> Weight;
+	fn swap_with_exact_target() -> Weight;
 }
 
 pub trait Trait: system::Trait {
-	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
 
 	/// Currency for transfer currencies
 	type Currency: MultiCurrencyExtended<Self::AccountId, CurrencyId = CurrencyId, Balance = Balance>;
 
-	/// CDP treasury for depositing additional liquidity reward to DEX
-	type CDPTreasury: CDPTreasury<Self::AccountId, Balance = Balance, CurrencyId = CurrencyId>;
-
-	/// Allowed trading currency type list, each currency type forms a trading
-	/// pair with the base currency
-	type EnabledCurrencyIds: Get<Vec<CurrencyId>>;
-
-	/// The base currency as the core currency in all trading pairs
-	type GetBaseCurrencyId: Get<CurrencyId>;
+	/// Allowed trading pair list.
+	type EnabledTradingPairs: Get<Vec<(CurrencyId, CurrencyId)>>;
 
 	/// Trading fee rate
-	type GetExchangeFee: Get<Rate>;
+	/// The first item of the tuple is the numerator of the fee rate, second
+	/// item is the denominator, fee_rate = numerator / denominator
+	type GetExchangeFee: Get<(u32, u32)>;
```

Comment:
why not using `Rate`?

Reply:
`target -> supply` calculation is the inverse of `supply -> target` calculation, in order to make the result of `target -> supply` can also meet `supply -> target`, we need add extra `1` after each internal `div` on `target -> supply` to correct the possible losses. However, the correction maybe make the result bigger, now we only have just 1 div on the end of `get_supply_amount`, and the final result is only 1 more at most.

These's a `div` in the `saturating_mul_int` of Rate, if use Rate, the formula also needs to be adjusted, and finally there will be 3 corrections. After 2 corrections, there are some other calculations, and the final result will have a large deviation. We have to minimize internal division operation.

Reply:
Should add something to comment, like use (u32, u32) over Rate to minimize internal division operation.